use home cache directories for PKGDIR and DISTDIR

stkw0 opened the pull request.

stkw0 committed:
use home cache directories for PKGDIR and DISTDIR

stkw0 committed:
mount portidr as ro in batch mode

In batch mode we don't drop into a shell. Therefore, it's safe to mount
it as read-only as no modifications would happen.

stkw0 commented:
@kuraga can you test if this changes helps with #237? I tried to test it with podman using a portage dir which had root:root and it worked for me.

It also mounts portage dir as RO if `--batch` is used.

nicolasbock reviewed (approved):
LGTM

kuraga commented:
@stkw0 , no, doesn't fix for me.

kuraga commented on ebuildtester/docker.py:
```diff
@@ -148,15 +149,23 @@ def _setup_container(self, docker_image):
+        pkgdir = "{}/packages".format(cache_dir)
         os.makedirs(pkgdir, exist_ok=True)
 
+        if options.OPTIONS.batch:
+            portdir_opt = "ro"
+        else:
+            portdir_opt = "rw"
```
```suggestion
        portdir_opt = "ro" if options.OPTIONS.batch else "rw"
```

stkw0 replied:
I thought about that but seemed less readable

kuraga commented on ebuildtester/docker.py:
```diff
@@ -167,9 +176,9 @@ def _create_container(self, docker_image, local_portage, overlays):
-               "--volume", "%s:/var/db/repos/gentoo" % local_portage,
-               "--volume", "%s:/var/cache/distfiles" % distdir,
-               "--volume", "%s:/var/cache/binpkgs" % pkgdir]
+               "--volume", f"{portdir}:/var/db/repos/gentoo:{portdir_opt}",
```
Note there are three types of string interpolation at the code of this patch.

kuraga commented:
> @kuraga can you test if this changes helps with #237? I tried to test it with podman using a portage dir which had root:root and it worked for me.

@stkw0 , no, it _doesn't_ fix for me.

UPDATE: _But_ it will fix if the host user is `root` _and_ the host's `{portdir}/**/*` are owned by `portage:portage`.

kuraga commented:
@stkw0 ,

the host user is `sasha`.

1. `/home/sasha/gentoo/*/**` are `sasha:sasha`: exactly #237 .

2. `/home/sasha/gentoo/*/**` are `portage:portage` (as the Handbook [says](https://wiki.gentoo.org/wiki/Handbook:AMD64/Portage/CustomTree#Creating_a_custom_ebuild_repository)):

```console
$ ebuildtester --portage-dir /home/sasha/gentoo --docker-command podman --docker-image docker.io/gentoo/stage3 --rm --batch --atom app-editors/nano
2025-07-30 13:53:24,876 - logging at /tmp/ebuildtester-app-editors-nano.log
2025-07-30 13:53:24,876 - *** please note that all necessary licenses will be accepted ***
2025-07-30 13:53:24,876 - creating container
2025-07-30 13:53:24,876 - creating docker container with: podman create --tty --cap-add CAP_SYS_ADMIN --cap-add CAP_MKNOD --cap-add CAP_NET_ADMIN --security-opt apparmor:unconfined --device /dev/fuse --workdir /root --volume /home/sasha/gentoo:/var/db/repos/gentoo:ro --volume /home/sasha/.cache/ebuildtester/distfiles:/var/cache/distfiles --volume /home/sasha/.cache/ebuildtester/packages:/var/cache/binpkgs docker.io/gentoo/stage3
2025-07-30 13:53:24,965 - container id 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:25,059 - setting Gentoo profile to default/linux/amd64/23.0
2025-07-30 13:53:25,059 - 48092a eselect profile set default/linux/amd64/23.0
2025-07-30 13:53:25,511 - 48092a (stderr): !!! Error: get_repo_path failed
2025-07-30 13:53:25,565 - running in container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:25,565 - failed command "eselect profile set default/linux/amd64/23.0"
2025-07-30 13:53:25,566 - no global USE flags given, skipping
2025-07-30 13:53:25,566 - tweaking portage settings
2025-07-30 13:53:25,566 - 48092a echo FEATURES=\"-sandbox -usersandbox userfetch\" >> /etc/portage/make.conf
2025-07-30 13:53:25,676 - 48092a echo MAKEOPTS=\"-j16\" >> /etc/portage/make.conf
2025-07-30 13:53:25,851 - 48092a echo "C.UTF-8 UTF-8" > /etc/locale.gen
2025-07-30 13:53:26,026 - enabling test feature for [Atom("app-editors/nano")]
2025-07-30 13:53:26,026 - 48092a mkdir -p /etc/portage/env
2025-07-30 13:53:26,190 - 48092a echo "app-editors/nano tester.conf" >> /etc/portage/package.env
2025-07-30 13:53:26,354 - 48092a echo "FEATURES=\"test splitdebug\"" > /etc/portage/env/tester.conf
2025-07-30 13:53:26,539 - unmasking [Atom("app-editors/nano")]
2025-07-30 13:53:26,539 - 48092a mkdir -p /etc/portage/package.accept_keywords
2025-07-30 13:53:26,675 - 48092a echo "app-editors/nano" ~amd64 >> /etc/portage/package.accept_keywords/testbuild
2025-07-30 13:53:26,780 - unmasking additional atoms
2025-07-30 13:53:26,780 - skipping update
2025-07-30 13:53:26,780 - skipping basic packages
2025-07-30 13:53:26,780 - setting gcc
2025-07-30 13:53:26,780 - summary
2025-07-30 13:53:26,780 - 48092a if [[ -d /etc/portage/package.accept_keywords ]]; then cat /etc/portage/package.accept_keywords/*; fi
2025-07-30 13:53:26,855 - 48092a (stdout): app-editors/nano ~amd64
2025-07-30 13:53:26,932 - 48092a if [[ -f /etc/portage/package.use/testbuild ]]; then cat /etc/portage/package.use/testbuild; fi
2025-07-30 13:53:27,058 - 48092a emerge --info
2025-07-30 13:53:27,411 - 48092a (stderr): Permission denied: '/var/db/repos/gentoo/eclass'
2025-07-30 13:53:27,493 - running in container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:27,493 - failed command "emerge --info"
2025-07-30 13:53:27,493 - 48092a qlop
2025-07-30 13:53:27,558 - 48092a (stderr): qlop: Could not open logfile '/var/log/emerge.log': No such file or directory
2025-07-30 13:53:27,600 - created container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:27,600 - 48092a echo emerge --verbose  --autounmask-write=y  --autounmask-license=y  --autounmask-continue=y  app-editors/nano --ask >> ~/.bash_history
2025-07-30 13:53:27,738 - emerge attempt 1 (of 5)
2025-07-30 13:53:27,738 - 48092a ['emerge', '--verbose ', '--autounmask-write=y ', '--autounmask-license=y ', '--autounmask-continue=y ', 'app-editors/nano']
2025-07-30 13:53:28,109 - 48092a (stderr): Permission denied: '/var/db/repos/gentoo/eclass'
2025-07-30 13:53:28,190 - running in container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:28,191 - failed command "['emerge', '--verbose ', '--autounmask-write=y ', '--autounmask-license=y ', '--autounmask-continue=y ', 'app-editors/nano']"
2025-07-30 13:53:28,191 - stopping container
48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:28,277 - deleting container
48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
```

Inside the container:

```console
0cd52e4e8810 ~ # ls -al /var/db/repos/gentoo/eclass
ls: cannot access '/var/db/repos/gentoo/eclass': Permission denied
0cd52e4e8810 ~ # ls -al /var/db/repos/gentoo       
ls: cannot open directory '/var/db/repos/gentoo': Permission denied
0cd52e4e8810 ~ # ls -al /var/db/repos       
total 12
drwxr-xr-x   1 portage portage 4096 Jul 30 10:56 .
drwxr-xr-x   1 root    root    4096 Jul 27 17:31 ..
drwxr-x--- 182 nobody  nobody  4096 Jul 29 20:27 gentoo
```

kuraga commented:
@stkw0 ,

the host user is `sasha`.

---

1. `/home/sasha/gentoo/*/**` are `sasha:sasha`: exactly #237:

```console
$ ebuildtester --portage-dir /home/sasha/gentoo --docker-command podman --docker-image docker.io/gentoo/stage3 --rm --atom app-editors/nano
<...>
2025-07-30 14:53:50,517 - creating docker container with: podman create --tty --cap-add CAP_SYS_ADMIN --cap-add CAP_MKNOD --cap-add CAP_NET_ADMIN --security-opt apparmor:unconfined --device /dev/fuse --workdir /root --volume /home/sasha/
gentoo:/var/db/repos/gentoo:rw --volume /home/sasha/.cache/ebuildtester/distfiles:/var/cache/distfiles --volume /home/sasha/.cache/ebuildtester/packages:/var/cache/binpkgs docker.io/gentoo/stage3                                          
WARN[0000] "/" is not a shared mount, this could cause issues or missing mounts with rootless containers                                                                                                                                     
<...>
2025-07-30 14:53:55,761 - 70613c (stdout): These are the packages that would be merged, in order:                     
2025-07-30 14:53:55,761 - 70613c (stdout):                                                                                                                                                                                                   
2025-07-30 14:53:55,896 - 70613c (stderr): /usr/lib/portage/python3.13/ebuild.sh: line 625: /var/db/repos/gentoo/app-alternatives/yacc/yacc-1-r2.ebuild: Permission denied                                                                   
2025-07-30 14:53:55,897 - 70613c (stderr):                                                                                                                                                                                                   
2025-07-30 14:53:55,897 - 70613c (stderr): Messages for package app-alternatives/yacc-1-r2::gentoo:                                                                                                                                          
2025-07-30 14:53:55,897 - 70613c (stderr):  * ERROR: app-alternatives/yacc-1-r2::gentoo failed (depend phase):                                                                                                                               
2025-07-30 14:53:55,899 - 70613c (stderr):  *   error sourcing ebuild                                                                                                                                                                        
2025-07-30 14:53:55,900 - 70613c (stderr):  *                                                                                                                                                                                                
2025-07-30 14:53:55,901 - 70613c (stderr):  * Call stack:                                                                                                                                                                                    
2025-07-30 14:53:55,905 - 70613c (stderr):  *   ebuild.sh, line 625:  Called die                                                                                                                                                             
2025-07-30 14:53:55,906 - 70613c (stderr):  * The specific snippet of code:                                                                                                                                                                  
2025-07-30 14:53:55,916 - 70613c (stderr):  *                           source "${EBUILD}" || die "error sourcing ebuild"                                                                                                                    
2025-07-30 14:53:55,918 - 70613c (stderr):  *                                                                                                                                                                                                
2025-07-30 14:53:55,919 - 70613c (stderr):  * If you need support, post the output of `emerge --info '=app-alternatives/yacc-1-r2::gentoo'`,                                                                                                 
2025-07-30 14:53:55,921 - 70613c (stderr):  * the complete build log and the output of `emerge -pqv '=app-alternatives/yacc-1-r2::gentoo'`.                                                                                                  
2025-07-30 14:53:55,924 - 70613c (stderr):  * Working directory: '/usr/lib/python3.13/site-packages'                                                                                                                                         
2025-07-30 14:53:55,925 - 70613c (stderr):  * S: '/var/tmp/portage/app-alternatives/yacc-1-r2/work/yacc-1'                                                                                                                                   
```

Inside the container:

```console
70613c81d60b ~ # ls -al /var/db/repos/gentoo/app-alternatives/yacc/                
total 16
drwxr-x---  2 root root 4096 Jul 11 17:42 .
drwxr-x--- 13 root root 4096 Jul 11 17:42 ..
-rw-r-----  1 root root  610 Jul 11 17:42 metadata.xml
-rw-r-----  1 root root 1137 Jul 11 17:42 yacc-1-r2.ebuild
```

---

2. As an addition, `/home/sasha/gentoo/*/**` are `portage:portage` (as the Handbook [says](https://wiki.gentoo.org/wiki/Handbook:AMD64/Portage/CustomTree#Creating_a_custom_ebuild_repository)):

```console
$ ebuildtester --portage-dir /home/sasha/gentoo --docker-command podman --docker-image docker.io/gentoo/stage3 --rm --batch --atom app-editors/nano
2025-07-30 13:53:24,876 - logging at /tmp/ebuildtester-app-editors-nano.log
2025-07-30 13:53:24,876 - *** please note that all necessary licenses will be accepted ***
2025-07-30 13:53:24,876 - creating container
2025-07-30 13:53:24,876 - creating docker container with: podman create --tty --cap-add CAP_SYS_ADMIN --cap-add CAP_MKNOD --cap-add CAP_NET_ADMIN --security-opt apparmor:unconfined --device /dev/fuse --workdir /root --volume /home/sasha/gentoo:/var/db/repos/gentoo:ro --volume /home/sasha/.cache/ebuildtester/distfiles:/var/cache/distfiles --volume /home/sasha/.cache/ebuildtester/packages:/var/cache/binpkgs docker.io/gentoo/stage3
2025-07-30 13:53:24,965 - container id 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:25,059 - setting Gentoo profile to default/linux/amd64/23.0
2025-07-30 13:53:25,059 - 48092a eselect profile set default/linux/amd64/23.0
2025-07-30 13:53:25,511 - 48092a (stderr): !!! Error: get_repo_path failed
2025-07-30 13:53:25,565 - running in container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:25,565 - failed command "eselect profile set default/linux/amd64/23.0"
2025-07-30 13:53:25,566 - no global USE flags given, skipping
2025-07-30 13:53:25,566 - tweaking portage settings
2025-07-30 13:53:25,566 - 48092a echo FEATURES=\"-sandbox -usersandbox userfetch\" >> /etc/portage/make.conf
2025-07-30 13:53:25,676 - 48092a echo MAKEOPTS=\"-j16\" >> /etc/portage/make.conf
2025-07-30 13:53:25,851 - 48092a echo "C.UTF-8 UTF-8" > /etc/locale.gen
2025-07-30 13:53:26,026 - enabling test feature for [Atom("app-editors/nano")]
2025-07-30 13:53:26,026 - 48092a mkdir -p /etc/portage/env
2025-07-30 13:53:26,190 - 48092a echo "app-editors/nano tester.conf" >> /etc/portage/package.env
2025-07-30 13:53:26,354 - 48092a echo "FEATURES=\"test splitdebug\"" > /etc/portage/env/tester.conf
2025-07-30 13:53:26,539 - unmasking [Atom("app-editors/nano")]
2025-07-30 13:53:26,539 - 48092a mkdir -p /etc/portage/package.accept_keywords
2025-07-30 13:53:26,675 - 48092a echo "app-editors/nano" ~amd64 >> /etc/portage/package.accept_keywords/testbuild
2025-07-30 13:53:26,780 - unmasking additional atoms
2025-07-30 13:53:26,780 - skipping update
2025-07-30 13:53:26,780 - skipping basic packages
2025-07-30 13:53:26,780 - setting gcc
2025-07-30 13:53:26,780 - summary
2025-07-30 13:53:26,780 - 48092a if [[ -d /etc/portage/package.accept_keywords ]]; then cat /etc/portage/package.accept_keywords/*; fi
2025-07-30 13:53:26,855 - 48092a (stdout): app-editors/nano ~amd64
2025-07-30 13:53:26,932 - 48092a if [[ -f /etc/portage/package.use/testbuild ]]; then cat /etc/portage/package.use/testbuild; fi
2025-07-30 13:53:27,058 - 48092a emerge --info
2025-07-30 13:53:27,411 - 48092a (stderr): Permission denied: '/var/db/repos/gentoo/eclass'
2025-07-30 13:53:27,493 - running in container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:27,493 - failed command "emerge --info"
2025-07-30 13:53:27,493 - 48092a qlop
2025-07-30 13:53:27,558 - 48092a (stderr): qlop: Could not open logfile '/var/log/emerge.log': No such file or directory
2025-07-30 13:53:27,600 - created container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:27,600 - 48092a echo emerge --verbose  --autounmask-write=y  --autounmask-license=y  --autounmask-continue=y  app-editors/nano --ask >> ~/.bash_history
2025-07-30 13:53:27,738 - emerge attempt 1 (of 5)
2025-07-30 13:53:27,738 - 48092a ['emerge', '--verbose ', '--autounmask-write=y ', '--autounmask-license=y ', '--autounmask-continue=y ', 'app-editors/nano']
2025-07-30 13:53:28,109 - 48092a (stderr): Permission denied: '/var/db/repos/gentoo/eclass'
2025-07-30 13:53:28,190 - running in container 48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:28,191 - failed command "['emerge', '--verbose ', '--autounmask-write=y ', '--autounmask-license=y ', '--autounmask-continue=y ', 'app-editors/nano']"
2025-07-30 13:53:28,191 - stopping container
48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
2025-07-30 13:53:28,277 - deleting container
48092a2750f402df72d8d38c2b7bdb284b52d0e13d7f5bf5918327294ed0f411
```

Inside the container:

```console
0cd52e4e8810 ~ # ls -al /var/db/repos/gentoo/eclass
ls: cannot access '/var/db/repos/gentoo/eclass': Permission denied
0cd52e4e8810 ~ # ls -al /var/db/repos/gentoo       
ls: cannot open directory '/var/db/repos/gentoo': Permission denied
0cd52e4e8810 ~ # ls -al /var/db/repos       
total 12
drwxr-xr-x   1 portage portage 4096 Jul 30 10:56 .
drwxr-xr-x   1 root    root    4096 Jul 27 17:31 ..
drwxr-x--- 182 nobody  nobody  4096 Jul 29 20:27 gentoo
0cd52e4e8810 ~ # ls -aln /var/db/repos       
total 12
drwxr-xr-x   1   250   250 4096 Jul 30 12:00 .
drwxr-xr-x   1     0     0 4096 Jul 27 17:31 ..
drwxr-x--- 182 65534 65534 4096 Jul 29 20:27 gentoo
```

Seems like it's expected: the host `portage:portage` (`250:250`) are mapped into the `65536, ...` range.

kuraga commented:
@stkw0 , the same.

```console
$ ls -aln gentoo/app-alternatives/yacc
total 16
drwxr-x---  2 1000 1000 4096 июл 30 17:42 .
drwxr-x--- 13 1000 1000 4096 июл 30 17:42 ..
-rw-r-----  1 1000 1000  610 июл 30 17:42 metadata.xml
-rw-r-----  1 1000 1000 1137 июл 30 17:42 yacc-1-r2.ebuild
$ id
uid=1000(sasha) gid=1000(sasha) groups=1000(sasha),10(wheel),27(video),78(kvm),100(users)
$ umask
0027
```

```console
e0e843fe254a ~ # ls -aln /var/db/repos/gentoo/app-alternatives/yacc
total 16
drwxr-x---  2 0 0 4096 Jul 30 14:42 .
drwxr-x--- 13 0 0 4096 Jul 30 14:42 ..
-rw-r-----  1 0 0  610 Jul 30 14:42 metadata.xml
-rw-r-----  1 0 0 1137 Jul 30 14:42 yacc-1-r2.ebuild
```

Yours?

stkw0 commented:
Also, did you run 
```
usermod --add-subuids 10000-75535 david
usermod --add-subgids 10000-75535 david 
podman system migrate 
```

kuraga commented:
@stkw0 , no.

```console
# cat /etc/subuid
sasha:100000:65536
# cat /etc/subgid
sasha:100000:65536
```

stkw0 commented:
The ro on --batch should be transparent to the user. It's more an implementation detail. But I would agree it should go on the Changelog of the next release

stkw0 committed:
use the calling user UID:GID as portage inside the container

The docker container creates the distfiles as "portage" user (UID 250)
in the host. This produces an error if then a container is launched as
podman. With this ugly hack we can get docker to fetch distfiles as the
calling user which avoids an error if then podman is used.

Closes: #237

kuraga commented:
@stkw0 , I'm sure it works. Thanks very much for the investigation.

But I _strictly_ vote to just add a requirement on permissions. Which would be up to the user.

But I don't know, which. Come back later.

kuraga commented:
Had an idea to add `root` to `portage` group in the container.
Doesn't work.

BTW, on the host:

```console
$ ls -aln /var/db/repos/gentoo/app-alternatives/yacc
total 20
drwxr-xr-x  2 0 0 4096 июл  8 23:07 .
drwxr-xr-x 13 0 0 4096 июл 19 16:07 ..
-rw-r--r--  1 0 0  597 мар 20 15:38 Manifest
-rw-r--r--  1 0 0  610 мар 20 15:38 metadata.xml
-rw-r--r--  1 0 0 1137 мар 20 15:38 yacc-1-r2.ebuild
```

So, the owner is `root`!

kuraga commented:
> > @stkw0 , What's your permission on an `.ebuild`-file?
> 
> ```
>  ❱ /bin/ls -aln app-alternatives/yacc
> total 8
> drwxr-xr-x 1 1000 1000   56 jul 10 02:05 .
> drwxr-xr-x 1 1000 1000  102 jul 10 02:05 ..
> -rw-r--r-- 1 1000 1000  610 mar 22 13:09 metadata.xml
> -rw-r--r-- 1 1000 1000 1137 jul 10 02:05 yacc-1-r2.ebuild
> ```

Another BTW:
it's not sufficient for me.
I need `o+x` permissions on `.ebuild`s, instead.

_Which_ user does `source` at `/usr/lib/portage/python3.13/ebuild.sh: line 625`?!

stkw0 commented:
> Had an idea to add `root` to `portage` group in the container.
> Doesn't work.

Yeah, if you don't use --userns=keep-id it would create a different UIDs which would mismatch

stkw0 commented:
Instead of the wrapper, reading the manual I found you can use an env var. for userns This way you could configure it once and forget about it. While writing this I realized we could just export that env var before calling "docker" (or podman). As docker don't have it implemented, it will just be ignored, making it compatible with both :D

kuraga commented:
> > > @stkw0 , What's your permission on an `.ebuild`-file?
> > 
> > ```
> >  ❱ /bin/ls -aln app-alternatives/yacc
> > total 8
> > drwxr-xr-x 1 1000 1000   56 jul 10 02:05 .
> > drwxr-xr-x 1 1000 1000  102 jul 10 02:05 ..
> > -rw-r--r-- 1 1000 1000  610 mar 22 13:09 metadata.xml
> > -rw-r--r-- 1 1000 1000 1137 jul 10 02:05 yacc-1-r2.ebuild
> > ```
> 
> Another BTW:
> it's not sufficient for me.
> I need `o+x` permissions on `.ebuild`s, instead.
> 
> _Which_ user does `source` at `/usr/lib/portage/python3.13/ebuild.sh: line 625`?!

@stkw0 , can it be some "non-superuser" `root` in the container?

In the container, I need `portage:portage` ownership or `o+x` permissions.

On the host, under `root`, I don't need.

Why? Thanks.

stkw0 commented:
No, a rootless podman container can not have an EUID=0 in the container, that would defear the purpose of a rootless container. It can have a root inside the container, but it's like a new user. I never used podman before, so I don't yet precisely how podman works under the hood

kuraga commented:
So yes, that's the case...

@stkw0 , does your patch require to start under `root`?

kuraga commented:
Well, does wrapper need `root`?

kuraga commented:
> ```
> /var/db/repos/gentoo
> ```

@stkw0 ,

1. If on the host file is owned by the user starting the container, who is it owned in the Docker container by?
2. If on the host file is owned by `root`, who is it owned in the Docker container by?

stkw0 commented:
Assuming a rootless podman with --userns=keep-id:uid=0
1. by root (the calling user is mapped to uid=0, root)
2. nobody (is expected)
3. nobody (I don't quite understand this)

Which is expected

stkw0 commented on ebuildtester/docker.py:
```diff
@@ -261,6 +261,7 @@ def _tweak_settings(self):
 
         # Avoid wasting time generating the whole set
         self.execute('echo "C.UTF-8 UTF-8" > /etc/locale.gen')
+        self.execute(f'sed -i "s/250/{os.getuid()}/g" /etc/passwd')
```
@nicolasbock can you check if this hack is acceptable? Thank you!

kuraga commented:
> Assuming a rootless podman with --userns=keep-id:uid=0
> 
> 1. by root (the calling user is mapped to uid=0, root)
> 2. nobody (is expected)
> 3. nobody (I don't quite understand this)
> 
> Which is expected

@stkw0 , I was asking about Docker. And without the hack.

But okay, I give up. The hack is required.

kuraga commented:
The hack works for me.

The only case I see without hack:

1. Start container by `root`.
2. We `chown` and `chmod` inside.
3. Or the user itself does p. 2.

kuraga commented:
@stkw0 , @nicolasbock , I see _three ways_ on Podman.
Don't know whether they work on Docker.

**The first one.**

The hack.

**The second one.**

1. Start container by `root`.
2. We `chown` and `chmod` inside.

**The third one.**

The user have to
1. Own the `${PORTAGE_DIR}` contents.
2. `chmod -R o+rx ${PORTAGE_DIR}` before the start.

I'd like the third, BTW ;)

kuraga commented:
> > But okay, I give up. The hack is required.
> 
> in docker is more simple. It's just the same than in the host. A file owned by the user, in Docker is the UID of the user, a file owned by root is root and owned by portage would appear as owned by portage.

If so, the only sufficient item on Docker (without the hack!) is:

> The user have to `chown -R portage:portage ${PORTAGE_DIR}` before start.

stkw0 commented:
Docker doesn't need to chown or anything. It just works. It's rootless podman and the ability to reuse distfiles/pkgfiles created by Docker in a rootless podman what is troublesome.

kuraga commented:
@stkw0 wrote in IRC chat:

> chown of the local portage dir would be a hassle if I then want to work in that repo as a user...

kuraga commented:
@stkw0 , I guess you mean "to work after `ebuildtester` finishes".

Well, yes. But we already pollute the repo directory.

Ok, on your choice.

stkw0 commented:
> Well, yes. But we already pollute the repo directory.

How so?

kuraga commented:
> > Well, yes. But we already pollute the repo directory.
> 
> How so?

In non-`--batch` mode we mount in `rw` and modify the directory.

kuraga commented:
BTW, I realize now that even for Podman `--userns` is global.
So we can't get `portage:portage` for one directory only in the container (just after the mount) until it would have owned by a separate user on the start.

https://github.com/nicolasbock/ebuildtester/pull/245#issuecomment-3146394827 has been updated.

stkw0 committed:
github workflows: switch from python 3.10 to 3.12

stkw0 committed:
github workflows: remove python 3.{8..10} and add 3.13

stkw0 committed:
functional workflow: add setuptools requirement

stkw0 committed:
switch from tomli to standard tomllib library

stkw0 committed:
switch to pyproject.toml

stkw0 committed:
workflows: comment out podman test

It seems it's uncompatible with github workflows.

stkw0 commented:
@nicolasbock Can you take a look at the latest 5 commits? If there is nothing bad I would like to merge this and some other small fixes and make a release.
